File: 0366-find-leaves-of-binary-tree/0366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> leaves;
    public TreeNode removeLeaves(TreeNode root) {
        if(root.left == null && root.right == null) {
            leaves.add(root.val);
            return null;
        }
        
        if(root.left!=null)root.left = removeLeaves(root.left);
        if(root.right!=null)root.right = removeLeaves(root.right);
        
        return root;
    }
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        
        while(true) {
            leaves = new ArrayList<>();
            if(root!=null) {
                root = removeLeaves(root);
            }else break;
            ans.add(leaves);
        }
        return ans;
    }
}

File: 0366-find-leaves-of-binary-tree/0366-find-leaves-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> leaves;
    TreeNode* removeLeaves(TreeNode* node) {
        //Base Case  for leaves.
        if(!node->left && !node->right){
            leaves.push_back(node->val);
            return nullptr;
        }
        
        //Left & right recursive call
        if(node->left) node->left = removeLeaves(node->left);
        if(node->right) node->right = removeLeaves(node->right);
        
        return node;
    }
    vector<vector<int>> findLeaves(TreeNode* root) {
        vector<vector<int>> ans;
        
        while(true) {
            leaves.clear();
            //if Tree exists or root still exists
            if(root) {
                root = removeLeaves(root);
            }
            //else tree is now null
            else break;
            ans.push_back(leaves);
        }
        return ans;
    }
};